### DISCUSSION TITLE
Why my neovim version is 0.9.5 in nixos

### DISCUSSION THREAD

**Discussion by solitudealma** [ayamir/nvimdots#1334] [Q&A]
even if i use unstable to install, my neovim still 0.9.5.
```nix
# This file defines overlays
{ inputs, ... }:
{
  # This one brings our custom packages from the 'pkgs' directory
  additions = final: prev: import ../pkgs { pkgs = final; };

  # This one contains whatever you want to overlay
  # You can change versions, add patches, set compilation flags, anything really.
  # https://nixos.wiki/wiki/Overlays
  modifications = final: prev: {
    # https://discourse.nixos.org/t/davinci-resolve-studio-install-issues/37699/44
    # https://theholytachanka.com/posts/setting-up-resolve/
    davinci-resolve = prev.davinci-resolve.override (old: {
      buildFHSEnv =
        a:
        (old.buildFHSEnv (
          a
          // {
            extraBwrapArgs = a.extraBwrapArgs ++ [ "--bind /run/opengl-driver/etc/OpenCL /etc/OpenCL" ];
          }
        ));
    });
    davinci-resolve-studio = prev.davinci-resolve-studio.override (old: {
      buildFHSEnv =
        a:
        (old.buildFHSEnv (
          a
          // {
            extraBwrapArgs = a.extraBwrapArgs ++ [ "--bind /run/opengl-driver/etc/OpenCL /etc/OpenCL" ];
          }
        ));
    });

    #linuxPackages_latest = prev.linuxPackages_latest.extend (_lpself: lpsuper: {
    #  mwprocapture = lpsuper.mwprocapture.overrideAttrs ( old: rec {
    #    pname = "mwprocapture";
    #    subVersion = "4390";
    #    version = "1.3.0.${subVersion}";
    #    src = prev.fetchurl {
    #      url = "https://www.magewell.com/files/drivers/ProCaptureForLinux_${subVersion}.tar.gz";
    #      sha256 = "sha256-a2cU7PYQh1KR5eeMhMNx2Sc3HHd7QvCG9+BoJyVPp1Y=";
    #    };
    #  });
    #});

    openasar = prev.openasar.overrideAttrs ( old: rec {
      pname = "openasar";
      version = "0-unstable-2024-06-30";
      src = prev.fetchFromGitHub {
        owner = "GooseMod";
        repo = "OpenAsar";
        rev = "5c875eb048e96543f1ec711fae522ace5e4a836c";
        hash = "sha256-dlf4X+2W2GfL2E46ZM35PqpcoKHoZ4fhroOCLpye1D0=";
      };
    });

    #wavebox = prev.wavebox.overrideAttrs ( old: rec {
    #  pname = "wavebox";
    #  version = "10.125.53-2";
    #  src = prev.fetchurl {
    #    url = "https://download.wavebox.app/stable/linux/tar/Wavebox_${version}.tar.gz";
    #    sha256 = "sha256-ymmo0SaE71wJe8i7qAiEvPdWIA5ePUfOS8JmhxFmQvI=";
    #  };
    #});
  };

  # When applied, the unstable nixpkgs set (declared in the flake inputs) will
  # be accessible through 'pkgs.unstable'
  unstable-packages = final: prev: {
    unstable = import inputs.nixpkgs-unstable {
      inherit (final) system;
      config.allowUnfree = true;
    };
  };
}
```

```nix
nixpkgs = {
    # You can add overlays here
    overlays = [
      # Add overlays your own flake exports (from overlays and pkgs dir):
      outputs.overlays.additions
      outputs.overlays.modifications
      outputs.overlays.unstable-packages
      # Add overlays exported from other flakes:
    ];
    # Configure your nixpkgs instance
    config = {
      allowUnfree = true;
    };
  };
```

```nix
nixpkgs.url = "github:nixos/nixpkgs/nixos-24.05";
# You can access packages and modules from different nixpkgs revs at the same time.
# See 'unstable-packages' overlay in 'overlays/default.nix'.
nixpkgs-unstable.url = "github:nixos/nixpkgs/nixos-unstable";
```

```
Error detected while processing /home/solitudealma/.config/nvim/init.lua:
E5113: Error while calling lua chunk: /home/solitudealma/.config/nvim/lua/core/global.lua:2: attempt to index field 'uv' (a nil value)
stack traceback:
        /home/solitudealma/.config/nvim/lua/core/global.lua:2: in main chunk
        [C]: in function 'require'
        /home/solitudealma/.config/nvim/lua/core/init.lua:2: in main chunk
        [C]: in function 'require'
        /home/solitudealma/.config/nvim/init.lua:2: in main chunk
Press ENTER or type command to continue
```

**Comment by charliie-dev** [comment]
it's not related to our config, u should check your nix config.

maybe @misumisumi can help if he got time.

**Reply by solitudealma** [reply]
thx，wait him to solve

**Comment by misumisumi** [comment]
It seems like neovim in nixpkgs 24.05 is still at version 0.95.
You need to declare `programs.neovim.package=pkgs.unstable-packages.neovim-unwrapped;` if you are using that overlay.

**Reply by solitudealma** [reply]
yes, because I found the nixo.org unstable is 0.10.0. so I use overlay to install. can't I do this? I'll try your method some time.

**Reply by misumisumi** [reply]
I did not recognize your overlay correctly.
I think the correct one is `programs.neovim.package=pkgs.unstable.neovim-unwrapped;`.
Did you already set this up?

**Reply by solitudealma** [reply]
wait, i am in windows. Start Linux now

**Reply by charliie-dev** [reply]
based on my experiences, it's better to use different name for different version of pkgs, even u use overlays or different versions of nixpkgs. u will get a more consistent experience.

**Reply by solitudealma** [reply]
> based on my experiences, it's better to use different name for different version of pkgs, even u use overlays or different versions of nixpkgs. u will get a more consistent experience.

I very much agree with your point of view, I did use unstable and 24.05 to distinguish. But I didn't expect to cover it directly in programs.neovim.package. I installed it directly using unstable.neovim on environment.packages. Although the words neovim 0,10,0 were indeed seen during the build process, the overwriting was successful.

**Reply by misumisumi** [reply]
>  I installed it directly using unstable.neovim on environment.packages. 

In Nixpkgs, there are often two levels of derivations:

- **Low-level derivations:** These focus on building the package itself. (like `neovim-unwrapped`)
- **High-level derivations:** These encapsulate the low-level derivation along with environment variables, configuration files, and other necessary components to create a fully functional environment or application. (like `neovim`)

Packages specifically declared in home-manager often specify options for generating higher-level packages.

**Reply by solitudealma** [reply]
This is understandable. Any configured's package option is used to override the installation package in the URL introduced in flake.nix or the default installation package. Unwrapped can be understood as not yet packaged. Using this type of package can usually get the latest commit.

I don’t know much about nix philosophy. Many questions can only be guessed by looking at other people’s configurations. hh. There are still many things to learn.

**Reply by charliie-dev** [reply]
u can check my nix/hm config as an example:
https://github.com/CharlesChiuGit/nix-hm-config/blob/a72332dc3876d62e6d1bf35052b1122dd5a32bb4/home.nix#L207-L246

do noted that i'm using nix/hm on non-NixOS distros, such as ubuntu and macos. it's a bit different from using home-manager as a NixOS module.

**Reply by solitudealma** [reply]
> u can check my nix/hm config as an example: https://github.com/CharlesChiuGit/nix-hm-config/blob/a72332dc3876d62e6d1bf35052b1122dd5a32bb4/home.nix#L207-L246
> 
> do noted that i'm using nix/hm on non-NixOS distros, such as ubuntu and macos. it's a bit different from using home-manager as a NixOS module.

When there was a problem with my nvimdots configuration, I also thought about this non-nixos installation method hhh. But since I use nix, I will take advantage of it. So I came here to disturb you. After all, I'm not in any rush to use it.
